### PR TITLE
Enable support for Openstack cloud.

### PR DESCRIPTION
**Pull request by arcprabh**
Signed-off-by: arcprabh <arcprabh@in.ibm.com>

### CODE REVIEW

File: run_kraken.py [hunk 1]
```diff
@@ -31,6 +32,8 @@ def get_node_scenario_object(node_scenario):
         return aws_node_scenarios()
     elif node_scenario['cloud_type'] == 'gcp':
         return gcp_node_scenarios()
+    elif node_scenario['cloud_type'] == 'openstackcloud':
```
Comment (paigerube14) [inline run_kraken.py:35]: Could we do ` 'openstack' in node_scenario['cloud_type']` instead, just to make the cloud type a little shorter

File: docs/node_scenarios.md [hunk 2]
```diff
@@ -32,6 +32,10 @@ After creating the service account you'll need to enable the account using the f
 
 Use 'generic' or do not add the 'cloud_type' key to your scenario if your cluster is not set up using one of the current supported cloud types
 
+#### OPENSTACK
+
+**NOTE**: For clusters with OPENSTACK Cloud, make sure [OPENSTACK CLI](pip3 install python-openstackclient) is installed and then create and source the [OPENSTACK RC file](https://docs.openstack.org/newton/user-guide/common/cli-set-environment-variables-using-openstack-rc.html) from the server where Kraken runs.
```
Comment (paigerube14) [inline docs/node_scenarios.md:37]: I think we can add the openstack python client to the requirements

Comment (paigerube14) [reply docs/node_scenarios.md:37]: Also, can we move this section up above line 31 so all the clouds are together and maybe add another line in-between the end of openstack and the current line 31.

**Comment by arcprabh** [comment]
New commit link is https://github.com/arcprabh/kraken/commit/92f81ed616b416c55ac602fedaa7ac98d9d4f1e2
Please review @paigerube14

File: kraken/node_actions/openstack_node_scenarios.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    # Node scenario to stop the node
+    def node_stop_scenario(self, instance_kill_count, node, timeout):
+        for _ in range(instance_kill_count):
+            try:
+                logging.info("Starting node_stop_scenario injection")
+                logging.info("Stopping the node %s " % (node))
+                openstack_node_ip = nodeaction.get_node_ip(node)
+                openstack_node_name = self.openstackcloud.get_openstack_nodename(openstack_node_ip)
+                self.openstackcloud.stop_instances(openstack_node_name)
+                self.openstackcloud.wait_until_running(openstack_node_name)
```
Comment (paigerube14) [inline kraken/node_actions/openstack_node_scenarios.py:93]: Should be `wait_until_stopped` instead of running

**Comment by paigerube14** [comment]
In the _get_openstack_nodename_ function I ran into a problem 


`server_list = runcommand.invoke("openstack server list | grep %s" % (os_node_ip))
        list_of_servers = server_list.split('\n')

        for item in list_of_servers:
            items = item.split("|")
            counter = 0
            for i in items:
                if i.strip() != "" and counter == 2:
                    node_name = i.strip()
                    counter += 1
                    continue
                item_list = i.split('=')
                if len(item_list) == 2 and item_list[-1].strip() == os_node_ip:
                    return node_name
                counter += 1`

**Comment by paigerube14** [comment]
In the get_openstack_nodename function I ran into a problem with getting the node name from the IP address. It seemed that in our OpenStack server list there were multiple instance matches when _greping_ for the IP address (see example response below) 

Ex.) runcommand.invoke("openstack server list | grep %s" % (os_node_ip))
 server list response: | 48e050e7-225b-4942-805e-f744f2bf3f62 | _ug-bootstrap_                                         | ACTIVE  | ugd2932022589-rm94d-openshift=192.168.3.159, 10.0.102.170                     | rhcos-42.80.20191010.0                      | m1.xlarge      |
| a2ed6f85-5a84-4249-9eb8-c029541ec464 | _qe-pr-osp2-hzmvf-worker-0-87jks_                                       | ACTIVE  | qe-pr-osp2-hzmvf-openshift=192.168.3.15                                       | qe-pr-osp2-hzmvf-rhcos                      | m1.large       |

So the response from the function was the 2 node names concatenated and then when it tried to stop the node the OpenStack client couldn’t find a node with name “ug-bootstrapqe-pr-osp2-hzmvf-worker-0-87jks”

I coded up something to get past that point. Feel free to rewrite if there’s a better way 
`server_list = runcommand.invoke("openstack server list | grep %s" % (os_node_ip)) list_of_servers = server_list.split('\n')
    for item in list_of_servers:
        items = item.split("|")
        counter = 0
        for i in items:
            if i.strip() != "" and counter == 2:
                node_name = i.strip()
                counter += 1
                continue
            item_list = i.split('=')
            if len(item_list) == 2 and item_list[-1].strip() == os_node_ip:
                return node_name
            counter += 1`

I’m not sure if this depends on the set up of an Openshift cluster, but for me the node name and the name of the open stack instance are the same so the get ip address and get open stack node name are redundant. If this is not the case for your set up please keep the functions; just wanted to call it out that that’s how our test environments are set up. 

Are you also able to add in the terminate node code when you get a chance. 

Looking great though! Thanks for the contributions and quick code changes already

File: kraken/node_actions/openstack_node_scenarios.py (hunk 3 above)
Comment (arcprabh) [reply kraken/node_actions/openstack_node_scenarios.py:93]: Good catch.  I will change it.

**Comment by paigerube14** [comment]
After changing those couple of things listed above I was able to succesfully start, stop and reboot nodes in Openstack :) 

Please be sure to run "tox" as well to correct all formatting issues.

**Comment by arcprabh** [comment]
> In the get_openstack_nodename function I ran into a problem with getting the node name from the IP address. It seemed that in our OpenStack server list there were multiple instance matches when _greping_ for the IP address (see example response below)
> 
> Ex.) runcommand.invoke("openstack server list | grep %s" % (os_node_ip))
> server list response: | 48e050e7-225b-4942-805e-f744f2bf3f62 | _ug-bootstrap_ | ACTIVE | ugd2932022589-rm94d-openshift=**192.168.3.15**9, 10.0.102.170 | rhcos-42.80.20191010.0 | m1.xlarge |
> | a2ed6f85-5a84-4249-9eb8-c029541ec464 | _qe-pr-osp2-hzmvf-worker-0-87jks_ | ACTIVE | qe-pr-osp2-hzmvf-openshift=**192.168.3.15** | qe-pr-osp2-hzmvf-rhcos | m1.large |
> 
> So the response from the function was the 2 node names concatenated and then when it tried to stop the node the OpenStack client couldn’t find a node with name “ug-bootstrapqe-pr-osp2-hzmvf-worker-0-87jks”
> 
> I coded up something to get past that point. Feel free to rewrite if there’s a better way
> 
> `
> server_list = runcommand.invoke("openstack server list | grep %s" % (os_node_ip))
> 
> ```
>  list_of_servers = server_list.split('\n')
> 
> for item in list_of_servers:
> 
>     items = item.split("|")
> 
>     counter = 0
> 
>     for i in items:
> 
>         if i.strip() != "" and counter == 2:
> 
>            node_name = i.strip()
> 
>             counter += 1
> 
>             continue
> 
>         item_list = i.split('=')
> 
>         if len(item_list) == 2 and item_list[-1].strip() == os_node_ip:
> 
>             return node_name
> 
>         counter += 1
> ```
> 
> `
> 
> I’m not sure if this depends on the set up of an Openshift cluster, but for me the node name and the name of the open stack instance are the same so the get ip address and get open stack node name are redundant. If this is not the case for your set up please keep the functions; just wanted to call it out that that’s how our test environments are set up.
> 
> Are you also able to add in the terminate node code when you get a chance.
> 
> Looking great though! Thanks for the contributions and quick code changes already

Tested this code and it works on my set up too.

**Comment by paigerube14** [comment]
LGTM

File: kraken/node_actions/openstack_node_scenarios.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+import logging
+import subprocess
+import requests
+import kraken.kubernetes.client as kubecli
+import kraken.invoke.command as runcommand
+import kraken.node_actions.common_node_functions as nodeaction
+from kraken.node_actions.abstract_node_scenarios import abstract_node_scenarios
+
+
+class OPENSTACKCLOUD:
```
Comment (pruan-rht) [inline kraken/node_actions/openstack_node_scenarios.py:12]: @paigerube14 I think python [PEP8](https://www.python.org/dev/peps/pep-0008/#class-names) states that class names should be CamelCase.

File: kraken/node_actions/openstack_node_scenarios.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                   node_name = i.strip()
+                   logging.info("Openstack node name is %s " % (node_name))
+                   counter += 1
+                   continue
+                item_list = i.split('=')
+                if len(item_list) == 2 and item_list[-1].strip() == os_node_ip:
+                   return node_name
+                counter += 1
+ 
+class openstack_node_scenarios(abstract_node_scenarios):
```
Comment (pruan-rht) [inline kraken/node_actions/openstack_node_scenarios.py:73]: ditto

**Comment by mffiedler** [comment]
@pruan-rht we can do a follow up PR if needed for clean up.  Merging to get the functionality in.

File: kraken/node_actions/openstack_node_scenarios.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+from kraken.node_actions.abstract_node_scenarios import abstract_node_scenarios
+
+
+class OPENSTACKCLOUD:
+    def __init__(self):
+        self.Wait = 30
+
+    # Start the node instance
+    def start_instances(self,node):
+        runcommand.invoke("openstack server start %s" % (node))
```
Comment (smalleni) [inline kraken/node_actions/openstack_node_scenarios.py:18]: Thought about actually using the openstack client instead of subprocessing commands?

**Pull request merged**